test/userController.test.js

```diff
@@ -94,3 +94,143 @@ describe('GET /api/v1/users/:user_id endpoint', () => {
     expect(response.status).toBe(404);
   });
 });
+
+describe('PATCH /api/v1/users/updateCurrentUser endpoint', () => {
+  it('should update the current user', async () => {
+    const signedCookie = await loginAndReturnCookie({
+      email: 'ava@ava.com',
+      password: 'secret',
+    });
+
+    const updatedUserData = {
+      email: 'new_email@example.com',
+      name: 'new_name',
+    };
+
+    const response = await request(app)
+      .patch('/api/v1/users/updateCurrentUser')
+      .send(updatedUserData)
+      .set('Cookie', [signedCookie]);
+
+    expect(response.status).toBe(200);
+    expect(response.body).toHaveProperty('user');
+    expect(response.body.user).toMatchObject(updatedUserData);
+  });
+
+  it('should return an error if required data is missing', async () => {
+    const signedCookie = await loginAndReturnCookie({
+      email: 'ava@ava.com',
+      password: 'secret',
+    });
+
+    const response = await request(app)
+      .patch('/api/v1/users/updateCurrentUser')
+      .set('Cookie', [signedCookie]);
+
+    expect(response.status).toBe(400);
+    expect(response.body).toHaveProperty('msg', 'Please provide all values');
+  });
+});
+
+describe('GET /api/v1/users/getCurrentUserWithPurchasedAlbums endpoint', () => {
+  it('should return the data of the currently authenticated user with purchased albums', async () => {
+    // Act: Log in and get a signed cookie
+    const signedCookie = await loginAndReturnCookie(testUserCredentials);
+
+    // Act: Get the current user's information with purchased albums using the authenticated cookie
+    const response = await request(app)
+      .get('/api/v1/users/getCurrentUserWithPurchasedAlbums')
```

Comment:
When I tried to run this test (line 136), I got the following error:

```
 FAIL  test/userController.test.js
  ● GET /api/v1/users/getCurrentUserWithPurchasedAlbums endpoint › should return the data of the currently authenticated user with purchased albums

    expect(received).toBe(expected) // Object.is equality

    Expected: 200
    Received: 404

      146 |
      147 |     // Assert: Check the response status and user information
    > 148 |     expect(response.status).toBe(200);
          |                             ^
      149 |     expect(response.body).toHaveProperty('user');
      150 |     expect(response.body.user).toHaveProperty('name', 'ava');
      151 |     expect(response.body.user).toHaveProperty('email', 'ava@ava.com');

      at Object.toBe (test/userController.test.js:148:29)

2023-09-19 09:14:09 [INFO]: POST /api/v1/auth/login 201 65.700 ms - 101
2023-09-19 09:14:09 [ERROR]: Cast to ObjectId failed for value "getCurrentUserWithPurchasedAlbums" (type string) at path "_id" for model "User" - CastError: Cast to ObjectId failed for value "getCurrentUserWithPurchasedAlbums" (type string) at path "_id" for model "User"
    at ObjectId.Object.<anonymous>.ObjectId.cast (/Users/aasante/dev/practicum2023/backend/node_modules/mongoose/lib/schema/objectid.js:250:11)
    at ObjectId.Object.<anonymous>.SchemaType.applySetters (/Users/aasante/dev/practicum2023/backend/node_modules/mongoose/lib/schematype.js:1220:12)
    at ObjectId.Object.<anonymous>.SchemaType.castForQuery (/Users/aasante/dev/practicum2023/backend/node_modules/mongoose/lib/schematype.js:1632:15)
    at cast (/Users/aasante/dev/practicum2023/backend/node_modules/mongoose/lib/cast.js:356:32)
    at model.Query.Object.<anonymous>.Query.cast (/Users/aasante/dev/practicum2023/backend/node_modules/mongoose/lib/query.js:4911:12)
    at model.Query.Object.<anonymous>.Query._castConditions (/Users/aasante/dev/practicum2023/backend/node_modules/mongoose/lib/query.js:2232:10)
    at model.Query._findOne (/Users/aasante/dev/practicum2023/backend/node_modules/mongoose/lib/query.js:2519:8)
    at model.Query.exec (/Users/aasante/dev/practicum2023/backend/node_modules/mongoose/lib/query.js:4430:28)
    at processTicksAndRejections (node:internal/process/task_queues:96:5)
    at getSingleUser (/Users/aasante/dev/practicum2023/backend/src/controllers/userController.js:29:16)
2023-09-19 09:14:09 [INFO]: GET /api/v1/users/getCurrentUserWithPurchasedAlbums 404 16.441 ms - 67
```

We can see thta the test tried to make a `GET` request to the URL `/api/v1/users/getCurrentUserWithPurchasedAlbums`. And in the second last line in that error stacktrace, we can see that the error occured in the `getSingleUser` function in the `userController.js` file on line 29: [https://github.com/Code-the-Dream-School/dd-prac-team4-back/blob/PS145-user-test-read-endpoints/src/controllers/userController.js#L29](https://github.com/Code-the-Dream-School/dd-prac-team4-back/blob/PS145-user-test-read-endpoints/src/controllers/userController.js#L29)

So my next question to myself would be, "Why did it call that controller function, instead of the `getCurrentUserWithPurchasedAlbums` handler?".

Let's take a look at the router file: [https://github.com/Code-the-Dream-School/dd-prac-team4-back/blob/PS145-user-test-read-endpoints/src/routes/userRoutes.js#L28C1-L30](https://github.com/Code-the-Dream-School/dd-prac-team4-back/blob/PS145-user-test-read-endpoints/src/routes/userRoutes.js#L28C1-L30)

We see on line 29, that the correct path to get a user with their albums is actually: `/api/v1/showMe/withAlbums`. So what happened here is that express saw that the path didn't match any of the routes with strict string matching, so it assumed that `getCurrentUserWithPurchasedAlbums` must have been a user id and instead called the route on line 41. And the error occurs when Mongo complains that the string `getCurrentUserWithPurchasedAlbums` is not a valid ObjectId type.